staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_width.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	"math"
+	"net/http"
+	"net/url"
+)
+
+func newListWidthEstimator(countFn objectCountGetterFunc) WidthEstimatorFunc {
+	estimator := &listWidthEstimator{
+		countFn: countFn,
+	}
+	return estimator.estimate
+}
+
+type listWidthEstimator struct {
+	countFn objectCountGetterFunc
+}
+
+func (e *listWidthEstimator) estimate(r *http.Request) Width {
+	requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+	if !ok {
+		// no RequestInfo should never happen, but to be on the safe side
+		// let's return minimumSeatsList
+		return Width{Seats: maximumSeats}
+	}
+
+	query := r.URL.Query()
+	listOptions := metav1.ListOptions{}
+	if err := metav1.Convert_url_Values_To_v1_ListOptions(&query, &listOptions, nil); err != nil {
+		klog.ErrorS(err, "Failed to convert options while calculating request width")
+		return Width{Seats: minimumSeats}
+	}
+
+	count := e.countFn.Get(key(requestInfo))
+	isListFromCache := !shouldListFromStorage(query, &listOptions)
+
+	// TODO: For resources that implement indexes at the watchcache level,
+	//  we need to adjust the cost accordingly
+	var estimatedObjectsToBeProcessed int64
+	switch {
+	case isListFromCache:
+		estimatedObjectsToBeProcessed = count
+	default:
+		// Even if a selector is specified and we may need to list and go over more objects from etcd
+		// to produce the result of size <limit>, each individual chunk will be of size at most <limit>.
+		// As a result. the width of the request should be computed based on <limit> and the actual
+		// cost of processing more elements will be hidden in the request processing latency.
+		estimatedObjectsToBeProcessed = listOptions.Limit
+	}
+
+	klog.Infof("estimatedObjectsToBeProcessed=%d count=%d, key=%s isListFromCache=%t", estimatedObjectsToBeProcessed, count, key(requestInfo), isListFromCache)
```

Comment:
This should by logged with my lower verbosity imho (V(4) or sth).

Reply:
oops, debug code, removing it.